pennylane/math/quantum.py

```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
     return vn_entropy_1 + vn_entropy_2 - vn_entropy_12
 
 
+# pylint: disable=too-many-arguments
+def vn_entanglement_entropy(
+    state, indices0, indices1, base=None, check_state=False, c_dtype="complex128"
+):
+    r"""Compute the Von Neumann entanglement entropy between two subsystems in a given state.
+
+    .. math::
+
+        S(\rho_A) = -\text{Tr}[\rho_A \log \rho_A] = -\text{Tr}[\rho_B \log \rho_B] = S(\rho_B)
+
+    where :math:`S` is the von Neumann entropy, and :math:`\rho_A = \text{Tr}_B[\rho_{AB}]` and
+    :math:`\rho_B = \text{Tr}_A[\rho_{AB}]` are the reduced density matrices for each partition.
+
+    The Von Neumann entanglement entropy is a measure of the degree of quantum entanglement between
+    two subsystems constituting a pure bipartite quantum state. The entropy of entanglement is the
+    Von Neumann entropy of the reduced density matrix for any of the subsystems. If it is non-zero,
+    it indicates the two subsystems are entangled.
+
+    Each state must be given as a density matrix. To find the mutual information given
+    a pure state, call :func:`~.math.dm_from_state_vector` first.
+
+    Args:
+        state (tensor_like): ``(2**N)`` state vector or ``(2**N, 2**N)`` density matrix.
+        indices0 (list[int]): List of indices in the first subsystem.
+        indices1 (list[int]): List of indices in the second subsystem.
+        base (float): Base for the logarithm. If None, the natural logarithm is used.
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.

```

Comment:
Why do we need to pass this rather than have it inferred from `state`?

Reply:
I am not sure, I used the same arguments as in `mutual_info`

Reply:
No problem, we can check in a Core standup.

Reply:
According to Romain, this is required by the old device API where it is possible to set the dtype on the device, so this was needed to match the device dtype.

Reply:
Thanks for checking! It's still not clear to me why that can't be inferred by the `state` :thinking: Can the `state` be a different dtype to what the device wants?

Reply:
I tried removing all `c_dtype` casting from the module and it didn't seem to cause any test failures, so I'm unsure if this is still necessary anymore. Although I don't think it's best to make such a change in this PR. For now I think it's reasonable to make the call signature of `vn_entanglement_entroy` consistent with all other math functions.